File: ruby/lib/random_math_operations.rb
# frozen_string_literal: true

class RandomMathOperations
  PRIMES = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97].freeze

  def initialize(seed: nil)
    @rng = seed ? Random.new(seed) : Random.new
  end

  def generate_random_odd_number
    value = @rng.rand(50) * 2 + 1
    value
  end

  def generate_random_even_number
    value = @rng.rand(51) * 2
    value
  end

  def generate_random_prime_candidate
    PRIMES[@rng.rand(PRIMES.length)]
  end

  def self.prime?(n)
    return false if n < 2
    return true if n < 4
    return false if n.even?

    (3..Math.sqrt(n).to_i).step(2).none? { |i| (n % i).zero? }
  end
end

File: elixir/lib/random_math_operations.ex
defmodule CornTest.RandomMathOperations do
  @moduledoc """
  Random mathematical operations with intentional flakiness.
  """

  @spec generate_random_odd_number() :: integer()
  def generate_random_odd_number do
    Enum.random(0..49) * 2 + 1
  end

  @spec generate_random_even_number() :: integer()
  def generate_random_even_number do
    Enum.random(0..50) * 2
  end

  @spec generate_random_prime_candidate() :: integer()
  def generate_random_prime_candidate do
    primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]
    Enum.random(primes)
  end

  @spec is_prime(integer()) :: boolean()
  def is_prime(n) when n <= 1, do: false
  def is_prime(n) when n <= 3, do: true
  def is_prime(n) when rem(n, 2) == 0 or rem(n, 3) == 0, do: false

  def is_prime(n) do
    max = trunc(:math.sqrt(n))

    5..max//6
    |> Enum.all?(fn i -> rem(n, i) != 0 and rem(n, i + 2) != 0 end)
  end
end
